src/vehicle_gui.cpp

```diff
@@ -1384,44 +1535,99 @@ void BaseVehicleListWindow::DrawVehicleListItems(VehicleID selected_vehicle, int
 
-		if (v->IsChainInDepot()) {
-			str = STR_BLUE_COMMA;
-		} else {
-			str = (v->age > v->max_age - DAYS_IN_LEAP_YEAR) ? STR_RED_COMMA : STR_BLACK_COMMA;
-		}
+				DrawVehicleImage(v, image_left, image_right, y + FONT_HEIGHT_SMALL - 1, selected_vehicle, EIT_IN_LIST, 0);
 
-		SetDParam(0, v->unitnumber);
-		DrawString(left, right, y + 2, str);
+				if (!v->name.empty()) {
+					/* The vehicle got a name so we will print it */
+					SetDParam(0, v->index);
+					DrawString(text_left, text_right, y, STR_TINY_BLACK_VEHICLE);
+				} else if (v->group_id != DEFAULT_GROUP) {
+					/* The vehicle has no name, but is member of a group, so print group name */
+					SetDParam(0, v->group_id);
+					DrawString(text_left, text_right, y, STR_TINY_GROUP, TC_BLACK);
+				}
+
+				if (show_orderlist) DrawSmallOrderList(v, orderlist_left, orderlist_right, y, v->cur_real_order_index);
+
+				StringID str;
+				if (v->IsChainInDepot()) {
+					str = STR_BLUE_COMMA;
+				} else {
+					str = (v->age > v->max_age - DAYS_IN_LEAP_YEAR) ? STR_RED_COMMA : STR_BLACK_COMMA;
+				}
+
+				SetDParam(0, v->unitnumber);
+				DrawString(left, right, y + 2, str);
+				break;
+			}
 
-		DrawVehicleProfitButton(v, vehicle_button_x, y + FONT_HEIGHT_NORMAL + 3);
+			case GB_SHARED_ORDERS:
+
+				assert(vehgroup.NumVehicles() > 0);
+
+				for (int i = 0; i < static_cast<int>(vehgroup.NumVehicles()); ++i) {
+					if (image_left + 8 * i >= image_right) break; // Break if there is no more space to draw any more vehicles anyway.
+					DrawVehicleImage(vehgroup.vehicles_begin[i], image_left + 8 * i, image_right, y + FONT_HEIGHT_SMALL - 1, selected_vehicle, EIT_IN_LIST, 0);
+				}
+
+				if (show_orderlist) DrawSmallOrderList((vehgroup.vehicles_begin[0])->GetFirstOrder(), orderlist_left, orderlist_right, y);
+
+				SetDParam(0, vehgroup.NumVehicles());
+				DrawString(left, right, y + 2, STR_BLACK_COMMA);
+				break;
+
+			default:
+				NOT_REACHED();
+		}
 
 		y += line_height;
 	}
 }
 
+void BaseVehicleListWindow::UpdateSortingFromGrouping()
+{
+	/* Set up sorting. Make the window-specific _sorting variable
+	 * point to the correct global _sorting struct so we are freed
+	 * from having conditionals during window operation */
+	switch (this->vli.vtype) {
+		case VEH_TRAIN:    this->sorting = &_sorting[this->grouping].train; break;
+		case VEH_ROAD:     this->sorting = &_sorting[this->grouping].roadveh; break;
+		case VEH_SHIP:     this->sorting = &_sorting[this->grouping].ship; break;
+		case VEH_AIRCRAFT: this->sorting = &_sorting[this->grouping].aircraft; break;
+		default: NOT_REACHED();
+	}
+	this->vehgroups.SetSortFuncs(this->GetVehicleSorterFuncs());
+	this->vehgroups.SetListing(*this->sorting);
+	this->vehgroups.ForceRebuild();
+	this->vehgroups.NeedResort();
+}
+
+void BaseVehicleListWindow::UpdateVehicleGroupBy(GroupBy group_by)
+{
+	if (this->grouping != group_by) {
+		*this->sorting = this->vehgroups.GetListing(); // save the old sorting
```

Comment:
This comment I do not understand. It states what it is doing, but not why. So it is confusing me, as now I am curious what it is designed to do :D Could you either make a comment that also explains the why a bit? Cheers!

Reply:
Added, hopefully it's clearer now.

Reply:
Yes! That is exactly what I would like to read in a comment :) Tnx!